pkg/metadata/manager.go

```diff
@@ -0,0 +1,404 @@
+package metadata
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"sync"
+	"sync/atomic"
+
+	"github.com/cilium/cilium/pkg/cgroups"
+	"github.com/cilium/cilium/pkg/defaults"
+	v1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	nodetypes "github.com/cilium/cilium/pkg/node/types"
+	"github.com/cilium/cilium/pkg/option"
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "metadata")
+	// example cgroup path in kubernetes environments
+	// /run/cilium/cgroupv2/kubepods/burstable/pod1858680e-b044-4fd5-9dd4-f137e30e2180/e275d1a37782ab30008aa3ae6666cccefe53b3a14a2ab5a8dc459939107c8c0
+	defaultCgroupBasePath = cgroups.GetCgroupRoot() + "/kubepods"
+	// example cgroup path in nested kubernetes environments like kind
+	// /run/cilium/cgroupv2/kubelet/kubepods/pod4841248b-fc2f-41f4-9981-a685bf840ab5/d8f227cc24940cfdce8d8e601f3b92242ac9661b0e83f0ea57fdea1cb6bc93ec
+	defaultNestedCgroupBasePath = cgroups.GetCgroupRoot() + "/kubelet" + "/kubepods"
+	cgroupBasePaths             = []string{defaultCgroupBasePath, defaultNestedCgroupBasePath}
+)
+
+// Manager maintains Kubernetes and low-level metadata (cgroup path and
+// cgroup id) for local pods and their containers. In order to do that, it defines
+// and implements callback functions that are called on Kubernetes pod watcher events.
+// It also exposes APIs to read the saved metadata.
+//
+// During initialization, the manager checks for a valid base cgroup path.
+// In case of environments using non-default cgroup base paths, manager will
+// fail to get a valid cgroup base path, and ignore all the subsequent pod events.
+type Manager struct {
+	// Map of pod metadata indexed by their ids
+	podMetadataById map[string]*podMetadata
+	// Map of container metadata indexed by their cgroup ids
+	containerMetadataByCgrpId map[uint64]*containerMetadata
+	// Set to the valid cgroup base path if found
+	templateCgroupBasePath string
+	// Buffered channel to receive pod events
+	podEvents chan podEvent
+	// Object to check cgroup base path
+	checkCgroupPath *sync.Once
+	// Flag to check if manager is enabled, and processing events
+	enabled atomic.Bool
+	// Channel to shut down manager
+	shutdown chan struct{}
+	// Interface to do file operations
+	fschecker fs
+	// Interface to do cgroups related operations
+	cgroupsChecker cgroup
+}
+
+type PodMetadata struct {
+	name      string
+	namespace string
+	ips       []v1.PodIP
+}
+
+const (
+	PodAddEvent = iota
+	PodUpdateEvent
+	PodDeleteEvent
+	PodMetadataEvent
+)
+
+func NewManager() *Manager {
+	m := &Manager{
+		podMetadataById:           make(map[string]*podMetadata),
+		containerMetadataByCgrpId: make(map[uint64]*containerMetadata),
+		podEvents:                 make(chan podEvent, 20),
```

Comment:
The channel size here should be a configurable parameter on the commandline (even if it's hidden). Ideally during testing we can even set it to be very low (perhaps even serial) in order to test for deadlocks in the case that there are too many events in the system to handle correctly.

Reply:
^^ Looking over the implementation the risk of deadlock looks pretty low to me, since this code isn't calling out to the rest of the daemon, but I think we should be trying to move towards having configurable channel sizes and testing with low channel sizes so that we aren't relying on having large empty channel buffers just to ensure progress is made in the agent. I wonder whether we should just have a hidden generic channel size flag for defaults on stuff like this or whether each channel should get independent flags :thinking:

Reply:
Hmm... Right, there is no risk of deadlock as all the calls are serialized via the channel. The buffer size merely allows callers to not block on this channel. 

> I wonder whether we should just have a hidden generic channel size flag for defaults on stuff like this or whether each channel should get independent flags

Interesting point, although in this particular case when would you need to change the size if we are not worried about deadlocks? In general, I could see some value in setting channel buffer size to a low value in unit tests.

Reply:
>  I could see some value in setting channel buffer size to a low value in unit tests.

Yeah, this point is the main thing I'm thinking about. If we reduce all of these channels to a small value, then we can ensure that the system still makes progress and doesn't livelock/deadlock if there are too many events going on.

Reply:
So I updated the channel size in unit tests as discussed above.